src/Compilers/CSharp/Portable/Binder/LockBinder.cs

```diff
@@ -73,8 +74,18 @@ internal override BoundStatement BindLockStatementParts(BindingDiagnosticBag dia
                     errorCode: ErrorCode.ERR_BadSpecialByRefLock);
             }
 
-            BoundStatement stmt = originalBinder.BindPossibleEmbeddedStatement(_syntax.Statement, diagnostics);
+            var needsFilterDiagnostics = isLockObjectBased && diagnostics.AccumulatesDiagnostics;
+            var bodyDiagnostics = needsFilterDiagnostics ? BindingDiagnosticBag.GetInstance(template: diagnostics) : diagnostics;
+
+            BoundStatement stmt = originalBinder.BindPossibleEmbeddedStatement(_syntax.Statement, bodyDiagnostics);
             Debug.Assert(this.Locals.IsDefaultOrEmpty);
+
+            if (needsFilterDiagnostics)
```

Comment:
>needsFilterDiagnostics [](http://example.com/codeflow?start=16&length=22)

Could you point me to the test that observes effect of this filtering?

Reply:
`LockTests.Yield` and `Yield_Async`